### PR TITLE
Refactor BitReader to contain explicit state (#1282)

### LINKED ISSUE #1282: Don't Interwine Bit and Byte Aligned Operations in `BitReader`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

`BitReader` contains a mixture of byte-aligned APIs such as `get_aligned` and `get_vlq_int` along with bit-aligned APIs such as `get_value` and `get_batch`. This makes the implementation kind of hard to follow, there is a buffer but this is only used for bit-level APIs, there is a bit_offset, but this is not used for some APIs, etc...

**Describe the solution you'd like**

Explicitly split the implementation of aligned APIs that operate on the byte stream from unaligned APIs that operate on the bit stream.

**Describe alternatives you've considered**

The current code is not incorrect and could remain unchanged

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #1282.

# Rationale for this change
 
I think it makes the logic easier to follow, although this may be a subjective assessment

# What changes are included in this PR?

Refactors BitReader to explicitly separate the aligned and unaligned operations

# Are there any user-facing changes?

No

### CODE REVIEW

File: parquet/src/util/bit_util.rs [hunk 1]
```diff
@@ -445,158 +445,307 @@ impl BitWriter {
...
+        let remaining_bits = self.state.remaining_bits();
+        if remaining_bits < needed_bits {
+            values_to_read = remaining_bits / bit_width;
+        }
+
+        let mut values_read = 0;
+
+        // First align bit offset to byte offset
+        if let BitReaderState::Unaligned(unaligned) = &mut self.state {
+            while values_to_read != values_read && (unaligned.bit_offset % 8) != 0 {
```
Comment (tustvold) [inline parquet/src/util/bit_util.rs:720]: There is a slight change here - we align to the byte not `u64`. I think this was unintentional, as the comment said `byte offset`.

File: parquet/src/util/bit_util.rs [hunk 2]
```diff
@@ -445,158 +445,307 @@ impl BitWriter {
...
+        to_read: usize,
+        bit_width: usize,
+    ) -> usize {
+        assert!(bit_width <= 32);
+
+        let mut values_read = 0;
+        let in_buf = &self.buffer.data()[self.byte_offset..];
+        assert!(in_buf.len() * 8 >= to_read * bit_width);
+
+        let mut in_ptr = in_buf as *const [u8] as *const u8 as *const u32;
```
Comment (tustvold) [inline parquet/src/util/bit_util.rs:548]: This block is simply moved, with some slightly more verbose naming

File: parquet/src/util/bit_util.rs [hunk 3]
```diff
@@ -445,158 +445,307 @@ impl BitWriter {
...
+    #[inline]
+    pub fn get_byte_offset(&self) -> usize {
+        self.state.get_byte_offset()
+    }
+
+    /// Reads a value of type `T` and of size `num_bits`.
+    ///
+    /// Returns `None` if there's not enough data available. `Some` otherwise.
+    pub fn get_value<T: FromBytes>(&mut self, num_bits: usize) -> Option<T> {
+        self.state.as_unaligned().get(num_bits)
```
Comment (sunchao) [inline parquet/src/util/bit_util.rs:691]: Not sure whether there's any performance implication here since it now requires some extra compare & jump instructions.

**Review by sunchao — COMMENTED** [review]
This looks correct, just not sure how much extra costs `as_aligned` and `as_unaligned` will incur.

File: parquet/src/util/bit_util.rs (hunk 3 above)
Comment (tustvold) [reply parquet/src/util/bit_util.rs:691]: Running the benchmarks if anything I see a teeny performance improvement, this is likely because whilst there is a branch here, it isn't typically called in a loop.

Although `DeltaBitPackDecoder` is an exception here, but that's what #1284 is for :smile:

**Comment by tustvold** [comment]
Going to get #1284 in first and then revisit this

**Comment by tustvold** [comment]
Going to close this as it wasn't needed for #1284
